examples/step-57/step-57.cc

```diff
@@ -0,0 +1,879 @@
+  //
+  // This function implements the Newton iteration with given tolerance, maximum number of iterations,
+  // and the number of mesh refinements to do. "is_initial_step" is the flag to tell us whether
+  // "setup_system" is necessary, and which part,
+  // system matrix or right hand side vector, should be assembled. If we do a line search,
+  // the right hand side is already assembled while checking the residual norm in the last iteration.
+  // Therefore, we just need to assemble the system matrix at the current iteration. The last
+  // argument "output_result" is whether output should be produced.
+
+  template <int dim>
+  void StationaryNavierStokes<dim>::newton_iteration(const double tolerance,
+                                                     const unsigned int max_iteration,
+                                                     const unsigned int n_refinements,
+                                                     bool  is_initial_step,
+                                                     bool  output_result)
+  {
+    double current_res;
+    double last_res;
+    bool   first_step = is_initial_step;
+
+    for (unsigned int refinement = 0; refinement < n_refinements+1; ++refinement)
+      {
+        unsigned int outer_iteration = 0;
+        last_res = 1.0;
+        current_res = 1.0;
+        std::cout << "*****************************************" << std::endl;
+        std::cout << "************  refinement = " << refinement << " ************ " << std::endl;
+        std::cout << "viscosity= " << viscosity << std::endl;
+        std::cout << "*****************************************" << std::endl;
+
+        while ((first_step || (current_res > tolerance)) && outer_iteration < max_iteration)
+          {
+            if (first_step)
+              {
+                setup_dof();
+                initialize_system();
+                evaluation_point = present_solution;
+                assemble_matrix(first_step);
+                assemble_rhs(first_step);
+                solve(first_step);
+                present_solution = newton_update;
+                nonzero_constraints.distribute(present_solution);
+                first_step = false;
+                evaluation_point = present_solution;
+                assemble_rhs(first_step);
+                current_res = system_rhs.l2_norm();
+                std::cout << "******************************" << std::endl;
+                std::cout << " The residual of initial guess is " << current_res << std::endl;
+                std::cout << " Initialization complete!  " << std::endl;
+                last_res = current_res;
+              }
+
+            else
+              {
+                evaluation_point = present_solution;
+                assemble_matrix(first_step);
+                if (outer_iteration == 0)
+                  assemble_rhs(first_step);
+                solve(first_step);
+
+                for (double alpha = 1.0; alpha > 1e-5; alpha *= 0.5)
```

Comment:
since you haven't described how line search works in the introduction, this would be the point where it's worthwhile adding a comment that discusses it at least briefly. It would be ok to refer to step-15 for more details.